pkg/ccl/streamingccl/streamingest/ingest_span_configs.go

```diff
@@ -0,0 +1,253 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+type spanConfigIngestor struct {
+	accessor                    *spanconfigkvaccessor.KVAccessor
+	bufferedUpdates             []spanconfig.Record
+	bufferedDeletes             []spanconfig.Target
+	lastBufferedSourceTimestamp hlc.Timestamp
+	session                     sqlliveness.Session
+	cutoverSignal               cutoverProvider
+	cutoverCh                   chan struct{}
+	group                       ctxgroup.Group
+	settings                    *cluster.Settings
+	client                      streamclient.Client
+	rekeyer                     *backupccl.KeyRewriter
+}
+
+func makeSpanConfigIngestor(
+	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
+	ingestionJob *jobs.Job,
+	sourceTenantID roachpb.TenantID,
+) (*spanConfigIngestor, error) {
+
+	accessor := spanconfigkvaccessor.New(
+		execCfg.DB,
+		execCfg.InternalDB.Executor(),
+		execCfg.Settings,
+		execCfg.Clock,
+		systemschema.SpanConfigurationsTableName.FQString(),
+		nil, /* knobs */
+	)
+
+	client, err := connectToActiveClient(ctx, ingestionJob, execCfg.InternalDB, streamclient.ForSpanConfigs())
+	if err != nil {
+		return nil, err
+	}
+
+	details := ingestionJob.Details().(jobspb.StreamIngestionDetails)
+
+	rekeyCfg := execinfrapb.TenantRekey{
+		OldID: sourceTenantID,
+		NewID: details.DestinationTenantID,
+	}
+	rekeyer, err := backupccl.MakeKeyRewriterFromRekeys(keys.SystemSQLCodec,
+		nil /* tableRekeys */, []execinfrapb.TenantRekey{rekeyCfg},
+		true /* restoreTenantFromStream */)
+	if err != nil {
+		return nil, err
+	}
+
+	cutoverProvider := &cutoverFromJobProgress{
+		jobID: ingestionJob.ID(),
+		db:    execCfg.InternalDB,
+	}
+	log.Infof(ctx, "initialized span config ingestor")
+	return &spanConfigIngestor{
+		accessor:      accessor,
+		settings:      execCfg.Settings,
+		session:       ingestionJob.Session(),
+		cutoverSignal: cutoverProvider,
+		cutoverCh:     make(chan struct{}),
+		group:         ctxgroup.WithContext(ctx),
+		client:        client,
+		rekeyer:       rekeyer,
+	}, nil
+}
+
+func (sc *spanConfigIngestor) close(ctx context.Context) {
+	if err := sc.client.Close(ctx); err != nil {
+		log.Errorf(ctx, "error on client.close(): %s", err)
+	}
+	if err := sc.group.Wait(); err != nil {
+		log.Errorf(ctx, "error on close(): %s", err)
+	}
+}
+
+func (sc *spanConfigIngestor) ingestSpanConfigs(
+	ctx context.Context, tenantName roachpb.TenantName,
+) error {
+	sub, err := sc.client.SetupSpanConfigsStream(ctx, tenantName)
+	if err != nil {
+		return err
+	}
+
+	g := ctxgroup.WithContext(ctx)
+	g.GoCtx(sub.Subscribe)
+	g.GoCtx(sc.checkForCutover)
```

Comment:
Do we need this checkForCutover?  Could we instead have the goroutine running the distsql flow shut us down when it exits, in the same way it shuts down the replanner?

Reply:
Good point. Refactored to remove this.